### PR TITLE
Update online beamspot to quadruplet-only and rescale track uncertainties

### PR DESCRIPTION
**Pull request by makortel**
This PR makes the following updates to the online beamspot applications making them more robust:
* rescale pixel track parameter uncertainties to match better the reality
  * the uncertainties come from phase0 parametrization, so a difference wrt. phase1 is expected
  * @rovere tuned the scale factor such that the pulls of pixel vertex position coordinates were correct
* do pixel tracks only from quadruplets
  * triplets produce too many fakes that introduce small biases in the beamspot position
* enlargen the tracking region in Z from +- 3 cm to +- 6 cm

Tested in CMSSW_9_2_X_2017-06-09-1100. Should have no impact on runTheMatrix workflows.

@rovere @VinInn @mtosi @JanFSchulte @dinardo @sikler

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by dinardo** [comment]
Dear Experts,
concerning the error rescale factor, as you might have seen in the BeamPixel Online DQM monitor there is already a vertex error scale factor: the so called "VxErrCorr" in the beampixel_dqm_sourceclient-live_cfg.py configuration file.

Is that number still valid, or given your recent modifications that number should be now "1"?

Many thanks and kind regards,
- Mauro.

**Comment by slava77** [comment]
@cmsbuild please test

it should work this time

**Comment by rovere** [comment]
@dinardo 
Ciao Mauro,
in fact after deploying this PR in the DQM Playback system we had improved results from the beam_dqm module but almost no results from your beampixel module.
Can you test this PR with this year's data and check why with this PR the situation is worst? Maybe the parameter you were referring to could play a role.

Thanks,
M.

File: RecoPixelVertexing/PixelTrackFitting/interface/PixelFitterByHelixProjections.h [hunk 1]
```diff
@@ -10,29 +10,21 @@
...
-  float pt(float curvature) const;
-  float zip(float d0, float phi_p, float curv, 
-    const GlobalPoint& pinner, const GlobalPoint& pouter) const;
-  double errZip2(float apt, float eta) const;
-  double errTip2(float apt, float eta) const;
-  */
-private:
   const edm::EventSetup *theES;
   const MagneticField *theField;
+  const bool thescaleErrorsForPhaseI;
```
Comment (slava77) [inline RecoPixelVertexing/PixelTrackFitting/interface/PixelFitterByHelixProjections.h]: how does an instance of this class know that it operates on a phase-1 input or a phase-1 environment?
it looks like ```ForPhase1``` can be omitted

Comment (slava77) [inline RecoPixelVertexing/PixelTrackFitting/interface/PixelFitterByHelixProjections.h]: actually, based on implementation it should probably be
thescaleErrorsForBPIX1

Comment (makortel) [inline RecoPixelVertexing/PixelTrackFitting/interface/PixelFitterByHelixProjections.h]: I can certainly rename the parameter and variable to `theScaleErrorsForPhase1BPix1` or similar (I'd keep "phase1" in the name as the tuning was done with phase1 pixel, and is not necessarily applicable to phase2 (and clarly not for phase0)).

File: RecoPixelVertexing/PixelTrackFitting/python/pixelFitterByHelixProjections_cfi.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import FWCore.ParameterSet.Config as cms
+from Configuration.Eras.Modifier_phase1Pixel_cff import phase1Pixel
+
+from RecoPixelVertexing.PixelTrackFitting.pixelFitterByHelixProjectionsDefault_cfi import pixelFitterByHelixProjectionsDefault
+
+pixelFitterByHelixProjections = pixelFitterByHelixProjectionsDefault.clone()
+
+phase1Pixel.toModify( pixelFitterByHelixProjections, scaleErrorsForPhaseI = True)
```
Comment (makortel) [inline RecoPixelVertexing/PixelTrackFitting/python/pixelFitterByHelixProjections_cfi.py]: @slava77 `PixelFitterByHelixProjections` knows whether it operates on phase1 detector or not from this era customization.

**Comment by slava77** [comment]
On 6/22/17 4:53 AM, Matti Kortelainen wrote:
> *@makortel* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In
> RecoPixelVertexing/PixelTrackFitting/interface/PixelFitterByHelixProjections.h
> <https://github.com/cms-sw/cmssw/pull/19267#discussion_r123489707>:
> 
>>    const edm::EventSetup *theES;
>    const MagneticField *theField;
> +  const bool thescaleErrorsForPhaseI;
> 
> I can certainly rename the parameter and variable to
> |theScaleErrorsForPhase1BPix1| or similar (I'd keep "phase1" in the name
> as the tuning was done with phase1 pixel, and is not necessarily
> applicable to phase2 (and clarly not for phase0)).

If you prefer to keep "phase1" in the name, please add appropriate
software protections
that confirm that this is applied on phase1
(e.g. someone accidentally applying it on phase-0  or phase-2 should get
an exception or other error).

Otherwise, this modifier in the name appears inappropriate to me.
One could just as well desire to apply the same (with perhaps a
different value in the scale)
 for phase-2 for whatever reason and will generate quite a bit of confusion.

**Comment by slava77** [comment]
On 6/22/17 4:54 AM, Matti Kortelainen wrote:
> @slava77 <https://github.com/slava77> |PixelFitterByHelixProjections|
> knows whether it operates on phase1 detector or not from this era
> customization.
> 

My point is that if this name is to be kept in the parameter,
it should be impossible to apply it (by mistake or intentionally)  for
any other era/modifier

**Comment by makortel** [comment]
@slava77 Right, as the scale factor is also given as a parameter there should be no need to have "phase1" in the prameter/variable name (it is the scale factor that is phase1-specific, not the scaling itself). Will change along you suggested (but I'd prefer to wait until you have completed the review).

**Comment by slava77** [comment]
On 6/22/17 6:29 AM, Matti Kortelainen wrote:
> @slava77 <https://github.com/slava77> Right, as the scale factor is also
> given as a parameter there should be no need to have "phase1" in the
> prameter/variable name (it is the scale factor that is phase1-specific,
> not the scaling itself). Will change along you suggested (but I'd prefer
> to wait until you have completed the review).
> 

I was done with the review yesterday.
Sorry for lack of an explicit message.

**Comment by dmitrijus** [comment]
+1

**Comment by slava77** [comment]
@makortel 
please clarify if you are planning to update this soon
Thank you.

**Comment by makortel** [comment]
I'll push an update later today. 

On 23 June 2017 18:31:01 CEST, Slava Krutelyov <notifications@github.com> wrote:
>@makortel 
>please clarify if you are planning to update this soon
>Thank you.

-- 
Sent from my mobile.

**Comment by makortel** [comment]
@cmsbuild, please test

Renamed the parameter/variable as suggested by @slava77.

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19267	ebacaf6
- changes needed for online beamspot setup do not affect the standard reco processing
- jenkins tests pass and comparisons with baseline show no differences

**Comment by slava77** [comment]
@makortel 
I guess now this also needs to be back-ported to 92X.

**Comment by makortel** [comment]
Backport to 92X is here #19438.

**Comment by davidlange6** [comment]
merge

**Pull request merged**